File: 75.sort-colors.go
/*
 * @lc app=leetcode id=75 lang=golang
 *
 * [75] Sort Colors
 */

// @lc code=start
func sortColors(nums []int) {
	if len(nums) == 0 {
		return
	}

	r, w, b := 0, 0, 0

	for _, num := range nums {
		if num == 0 {
			nums[b] = 2
			b++
			nums[w] = 1
			w++
			nums[r] = 0
			r++
		} else if num == 1 {
			nums[b] = 2
			b++
			nums[w] = 1
			w++
		} else if num == 2 {
			b++
		}
	}
}

// @lc code=end

File: 75.sort-colors.cpp
/*
 * @lc app=leetcode id=75 lang=cpp
 *
 * [75] Sort Colors
 */

// @lc code=start
class Solution {
 public:
  void sortColors(vector<int>& nums) {
    int n = nums.size();

    int l = 0, h = n - 1;
    int cur = 0;

    while (cur <= h) {
      if (nums[cur] == 0) {
        swap(nums[l++], nums[cur++]);
      } else if (nums[cur] == 1) {
        cur++;
      } else {
        swap(nums[h--], nums[cur]);
      }
    }
  }
};
// @lc code=end

